lib/view_model/deserialize_context.rb

```diff
@@ -1,15 +1,29 @@
 class ViewModel
   class DeserializeContext
-    attr_accessor :updated_associations, :parent_ref
+    attr_accessor :updated_associations, :parent_context, :parent_viewmodel
 
-    def initialize(parent_ref: nil)
-      self.parent_ref = parent_ref
+    def initialize(*)
     end
 
-    def for_child(parent)
+    def parent(idx = 0)
+      if idx == 0
+        parent_viewmodel
+      else
+        parent_context&.parent(idx - 1)
+      end
+    end
+
+    def parent_ref(idx = 0)
+      parent(idx)&.to_reference
+    end
+
+    def for_child(parent_viewmodel)
       self.dup.tap do |copy|
-        copy.parent_ref = parent.to_reference
+        copy.parent_context   = self
+        copy.parent_viewmodel = parent_viewmodel
       end
     end
+
+    private :parent_context, :parent_viewmodel
```

Comment:
Awkward to leave the writers public. Is it worth making a method with a different name `initialize_descendant` to make it more obvious?